### PR TITLE
[Ipc4] add fw panic notification support

### PR DESCRIPTION
**Pull request by RanderWang**
Send fw panic notification message to host.

### CODE REVIEW

File: src/ipc/ipc4/handler.c [hunk 1]
```diff
@@ -1063,6 +1063,16 @@ void ipc_boot_complete_msg(struct ipc_cmd_hdr *header, uint32_t data)
 	header->ext = 0;
 }
 
+void ipc_send_panic_notification(void)
+{
+	msg_notify.header = SOF_IPC4_NOTIF_HEADER(SOF_IPC4_EXCEPTION_CAUGHT);
```
Comment (lyakh) [inline src/ipc/ipc4/handler.c:1068]: `msg_notify` only exists if `CONFIG_LOG_BACKEND_ADSP_MTRACE` is selected

File: src/ipc/ipc4/handler.c [hunk 2]
```diff
@@ -1063,6 +1063,16 @@ void ipc_boot_complete_msg(struct ipc_cmd_hdr *header, uint32_t data)
...
+	msg_notify.extension = cpu_get_id();
+	msg_notify.tx_size = 0;
+	list_init(&msg_notify.list);
+
+	ipc_msg_send_direct(&msg_notify, NULL);
```
Comment (lyakh) [inline src/ipc/ipc4/handler.c:1073]: I see that currently `msg_notify.tx_data` is never set, but if it's ever set to `!= NULL`, then the `msg->tx_data != data` inside `ipc_msg_send_direct()` (and `ipc_msg_send()`) will return `true` and they will try to copy from `NULL`. It's not a very robust API...

File: src/platform/library/include/platform/platform.h [hunk 3]
```diff
@@ -68,6 +68,8 @@ static inline void platform_wait_for_interrupt(int level) {}
 
 static inline int ipc_platform_send_msg(const struct ipc_msg *msg) { return 0; }
 
+static inline void ipc_platform_send_msg_direct(const struct ipc_msg *msg) { }
```
Comment (btian1) [inline src/platform/library/include/platform/platform.h:71]: seems there is a space in { }, no need?

File: src/ipc/ipc4/handler.c (hunk 1 above)
Comment (RanderWang) [reply src/ipc/ipc4/handler.c:1068]: Can you share more information ? I don't get it . Thanks!

File: src/ipc/ipc-common.c [hunk 4]
```diff
@@ -178,6 +178,26 @@ static void schedule_ipc_worker(void)
...
+void ipc_msg_send_direct(struct ipc_msg *msg, void *data)
+{
+	struct ipc *ipc = ipc_get();
+	k_spinlock_key_t key;
+	int ret;
+
+	key = k_spin_lock(&ipc->lock);
+
+	/* copy mailbox data to message if not already copied */
+	if ((msg->tx_size > 0 && msg->tx_size < SOF_IPC_MSG_MAX_SIZE) &&
```
[the remark this answers is not in the document]
Comment (RanderWang) [reply src/ipc/ipc-common.c:190]: sure, thanks

File: src/platform/library/include/platform/platform.h (hunk 3 above)
Comment (RanderWang) [reply src/platform/library/include/platform/platform.h:71]: yes, confirmed by other file. I will remove it

**Review by kv2019i — COMMENTED** [review]
Thanks @RanderWang for taking this forward. Some comments inline.

File: src/ipc/ipc4/handler.c (hunk 2 above)
Comment (RanderWang) [reply src/ipc/ipc4/handler.c:1074]: @lyakh  I set msg_notify.tx_data to NULL. Thanks!  Please comment if you have more ideas about it.

File: src/drivers/amd/rembrandt/ipc.c [hunk 5]
```diff
@@ -210,6 +210,11 @@ int ipc_platform_send_msg(const struct ipc_msg *msg)
 	return ret;
 }
 
+void ipc_platform_send_msg_direct(const struct ipc_msg *msg)
```
[the remark this answers is not in the document]
Comment (RanderWang) [reply src/drivers/amd/rembrandt/ipc.c:213]: sure, thanks

File: src/ipc/ipc-common.c [hunk 6]
```diff
@@ -178,6 +178,26 @@ static void schedule_ipc_worker(void)
...
+void ipc_msg_send_direct(struct ipc_msg *msg, void *data)
+{
+	struct ipc *ipc = ipc_get();
+	k_spinlock_key_t key;
+	int ret;
+
+	key = k_spin_lock(&ipc->lock);
+
+	/* copy mailbox data to message if not already copied */
+	if (msg->tx_size > 0 && msg->tx_size < SOF_IPC_MSG_MAX_SIZE &&
```
Comment (lyakh) [inline src/ipc/ipc-common.c:190]: actually, looking at it: shouldn't the second comparison be `<=`? "maximum size" usually is allowed too?

**Review by kv2019i — APPROVED** [review]
Thanks @RanderWang !

**Comment by kv2019i** [comment]
Known fails in Intel device tests https://sof-ci.01.org/sofpr/PR7597/build7839/devicetest/index.html https://sof-ci.01.org/sofpr/PR7597/build7838/devicetest/index.html . Zephyr upstream fails are known as well. Checkpatch warning we can live with. Proceeding with merge.

**Pull request merged**